pop_tools/budget.py

```diff
@@ -0,0 +1,328 @@
+    lmix_zonal = _compute_horizontal_divergence(ds.HDIFE, mask, direction='zonal')
+    lmix_merid = _compute_horizontal_divergence(ds.HDIFN, mask, direction='meridional')
+    lmix_B = ds.HDIFB.where(mask)
+    lmix_B = _compute_vertical_divergence(lmix_B)
+
+    # Sum all lateral mixing components
+    lmix = (lmix_merid + lmix_zonal + lmix_B).rename('lmix').sum('z_t')
+    lmix = _convert_units(lmix)
+    lmix.attrs['long_name'] = 'lateral mixing'
+    return lmix.load()
+
+
+def _compute_vertical_advection(ds, grid, mask, kmax=None):
+    """Compute vertical advection (WT)"""
+    print('Computing vertical advection...')
+    ds = ds['WT']
+    if kmax is not None:
+        # Need one level below max depth to compute divergence into bottom layer.
+        ds = ds.isel(z_t=slice(0, kmax + 2))
+    ds = ds.groupby('time.year').mean('time').rename({'year': 'time'})
+    ds = ds.where(mask)
+    if kmax is not None:
+        ds = _convert_to_tendency(ds, grid.vol, kmax=kmax + 1)
+    else:
+        ds = _convert_to_tendency(ds, grid.vol)
+
+    # Compute divergence of vertical advection.
+    vadv = (ds.shift(z_t=-1).fillna(0) - ds).isel(z_t=slice(0, -1))
+    vadv = vadv.sum('z_t').rename('vadv')
+    vadv = _convert_units(vadv)
+    vadv.attrs['long_name'] = 'vertical advection'
+    return vadv.load()
+
+
+def _compute_vertical_mixing(ds, grid, mask, kmax=None):
+    """Compute contribution from vertical mixing."""
+    print('Computing vertical mixing...')
+    ds = ds[['DIA_IMPVF', 'KPP_SRC']]
+    if kmax is not None:
+        ds = ds.isel(z_t=slice(0, kmax + 1))
+    ds = ds.groupby('time.year').mean('time').rename({'year': 'time'})
+    ds = ds.where(mask)
+    # Only need to flip sign of DIA_IMPVF.
+    ds['DIA_IMPVF'] = _convert_to_tendency(ds['DIA_IMPVF'], grid.area, sign=-1, kmax=kmax)
+    ds['KPP_SRC'] = _convert_to_tendency(ds['KPP_SRC'], grid.vol, kmax=kmax)
+
+    # Compute divergence of diapycnal mixing
+    diadiff = _compute_vertical_divergence(ds.DIA_IMPVF)
+
+    # Sum KPP and diapycnal mixing to create vmix term.
+    vmix = (ds.KPP_SRC + diadiff).rename('vmix').sum('z_t')
+    vmix = _convert_units(vmix)
+    vmix.attrs['long_name'] = 'vertical mixing'
+    return vmix.load()
+
+
+def _compute_SMS(ds, grid, mask, kmax=None):
+    """Compute SMS term from biology."""
+    print('Computing source/sink...')
+    ds = ds['SMS']
+    if (kmax is not None) and ('z_t' in ds.dims):
+        ds = ds.isel(z_t=slice(0, kmax + 1))
+    ds = ds.where(mask)
+    # Commonly output by POP as Jint_100m. So in those cases there won't be a
+    # depth-resolved dimension.
+    if 'z_t' in ds.dims:
+        ds = _convert_to_tendency(ds, grid.vol, kmax=kmax).rename('sms').sum('z_t')
+    else:
+        ds = _convert_to_tendency(ds, grid.area).rename('sms')
+    ds = _convert_units(ds)
+    ds = ds.groupby('time.year').mean('time').rename({'year': 'time'})
+    ds.attrs['long_name'] = 'source/sink'
+    return ds.load()
```

Comment:
Is it required to have eager execution here? What are the implications of having `ds` computed lazily?

Reply:
You bring up an interesting point. We could just return a completely lazy `ds` object back to the user to run `.compute()` themselves. Then they don't have to sit on a 5min or so idle computation. Perhaps they want to set up the task graph and save it out lazy to compute later? What are your thoughts here? Could have a `compute` keyword and either have the final return be `return reg_budget.load()` or `return reg_budget`, the latter of which would return a dask object.